src/config.rs

```diff
@@ -0,0 +1,98 @@
+//! The `config` module provides a generic mechanism for loading and managing
+//! request-scoped configuration.
+//!
+//! # Design Overview
+//!
+//! This module is centered around two abstractions:
+//!
+//! - The [`RequestConfigValue`] trait, used to associate a config key type with its value type.
+//! - The [`RequestConfig`] struct, which wraps an optional value of the type linked via [`RequestConfigValue`].
+//!
+//! Under the hood, the [`RequestConfig`] struct holds a single value for the associated config type.
+//! This value can be conveniently accessed, inserted, or mutated using [`http::Extensions`],
+//! enabling type-safe configuration storage and retrieval on a per-request basis.
+//!
+//! # Motivation
+//!
+//! The key design benefit is the ability to store multiple config types—potentially even with the same
+//! value type (e.g., [`Duration`])—without code duplication or ambiguity. By leveraging trait association,
+//! each config key is distinct at the type level, while code for storage and access remains totally generic.
+//!
+//! # Usage
+//!
+//! Implement [`RequestConfigValue`] for any marker type you wish to use as a config key,
+//! specifying the associated value type. Then use [`RequestConfig<T>`] in [`Extensions`]
+//! to set or retrieve config values for each key type in a uniform way.
+
+use std::time::Duration;
+
+use http::Extensions;
+
+/// This trait is empty and is only used to associate a configuration key type with its
+/// corresponding value type.
+pub(crate) trait RequestConfigValue: Copy + Clone + 'static {
+    type Value: Clone + Send + Sync + 'static;
+}
+
+/// RequestConfig carries a request-scoped configuration value.
+#[derive(Clone, Copy, Debug)]
+pub(crate) struct RequestConfig<T: RequestConfigValue>(Option<T::Value>);
+
+impl<T: RequestConfigValue> Default for RequestConfig<T> {
+    fn default() -> Self {
+        RequestConfig(None)
+    }
+}
+
+impl<T> RequestConfig<T>
+where
+    T: RequestConfigValue,
+{
+    pub(crate) fn new(v: Option<T::Value>) -> Self {
+        RequestConfig(v)
+    }
+
+    pub(crate) fn fetch<'client, 'request>(
+        &'client self,
+        ext: &'request Extensions,
+    ) -> Option<&'request T::Value>
+    where
+        'client: 'request,
+    {
+        ext.get::<RequestConfig<T>>()
+            .and_then(|v| v.0.as_ref())
+            .or(self.0.as_ref())
+    }
+
+    pub(crate) fn fetch_without_request(&self) -> Option<&T::Value> {
```

Comment:
I don't like this API but it's needed by `ClientRef::fmt_fields`. Can we remove it by implement `Debug` for `RequestConfig` instead of fetching it's raw value?

Reply:
Fixed by providing `fmt_as_field`.